table.cc

```diff
@@ -675,6 +675,7 @@ table::rebuild_sstable_list(const std::vector<sstables::shared_sstable>& new_sst
         if (!s.contains(tab)) {
             new_sstable_list.insert(tab);
         }
+        seastar::thread::maybe_yield();
```

Comment:
This function is always called on behalf of compaction, so it's guaranteed that it will always run in a thread. row_cache::invalidate() request that the the updater, i.e. rebuild_sstable_list(), provide strong exception guarantees, but rebuild_sstable_list() never did so, we even have a FIXME for that. with regards to its tolerance to yield, it looks like correctness will not be affected by the yield, but I'd like to have a confirmation from @tgrabiec about this.

Reply:
external_updater is not guaranteed to run inside a seastar thread. If it does, it's by chance.

Reply:
we guarantee that in this scenario that the updater is always called in a thread though. I'd like to confirm from you that yielding in the external updater will not violate requirements/correctness of invalidate().

Reply:
Even if row_cache::invalidate() is called in a thread, external_updater may not be called in a thread.

Reply:
it it ok if we change invalidate() to always call external updater in a thread? or is there any reason not to?

Reply:
I think it's ok

Reply:
How? It looks like external_updater is not called in a thread:

```c++
future<> row_cache::do_update(row_cache::external_updater eu, row_cache::internal_updater iu) noexcept {
    return futurize_invoke([this] {
        return get_units(_update_sem, 1);
    }).then([this, eu = std::move(eu), iu = std::move(iu)] (auto permit) mutable {
        auto pos = dht::ring_position::min();
        eu();
```

Reply:
@avikivity I deleted my comment by mistake "is it okay if we change do_update() to always call eu() in a thread?" which Tomek replied "I think it's ok" to. Sorry about that. Also, it may happen that external_updater is called in a thread if get_units() returns a ready future, which is why unit tests succeeded but any more elaborate testing would certainly fail. I plan to patch do_update() for v2.

Reply:
Maybe it's better to make external_updater return a future so we don't need to rely on hidden knowledge about what is called in a thread and what isn't.

Reply:
@avikivity Returning a future from external_updater won't work because cache needs to execute something atomically with external_updater updating the sstable list. If we run it in a thread, it works if external_updater does this as the last step. That's a bit shady.

Reply:
How can it be atomic with "eu" if "eu" is non-atomic? I guess it needs to be atomic with "the last part of eu", which happens to be updating the sstable list.

This must be made part of the interface, it is too tricky.

Perhaps

```c++
class external_updater_impl {
public:
    virtual ~external_updater_impl() {}
    future<> prepare() { return make_ready_future<>(); }
    virtual void execute() = 0;
};

using external_updater = std::unique_ptr<external_updater_impl>;

template <std::invocable_r<void> Execute>
external_updater
make_simple_external_updater(Execute func) {
    class thunk : public external_updater_impl {
         Execute _func;
    public:
          virtual void execute() override {
               _func();
          }
    };
}
```

We can run the prepare part first, then execute() atomically with the cache thing.

Reply:
> How can it be atomic with "eu" if "eu" is non-atomic? I guess it needs to be atomic with "the last part of eu", which happens to be updating the sstable list.

Yes, I think that's what I wrote.

> 
> This must be made part of the interface, it is too tricky.
> 
> Perhaps
> 
> ```c++
> class external_updater_impl {
> public:
>     virtual ~external_updater_impl() {}
>     virtual future<> prepare() { return make_ready_future<>(); }
>     virtual void execute() = 0;
> };
> 
> using external_updater = std::unique_ptr<external_updater_impl>;
> 
> template <std::invocable_r<void> Execute>
> external_updater
> make_simple_external_updater(Execute func) {
>     class thunk : public external_updater_impl {
>          Execute _func;
>     public:
>          virtual void execute() override {
>               _func();
>          }
>     };
>     return std::make_unique<thunk>(std::move(func));
> }
> ```
> 
> We can run the prepare part first, then execute() atomically with the cache thing.

Yes

Reply:
> > How can it be atomic with "eu" if "eu" is non-atomic? I guess it needs to be atomic with "the last part of eu", which happens to be updating the sstable list.
> 
> Yes, I think that's what I wrote.

Yes, I had to explain it to myself.

Reply:
> How can it be atomic with "eu" if "eu" is non-atomic? I guess it needs to be atomic with "the last part of eu", which happens to be updating the sstable list.
> 
> This must be made part of the interface, it is too tricky.
> 
> Perhaps
> 
> ```c++
> class external_updater_impl {
> public:
>     virtual ~external_updater_impl() {}
>     virtual future<> prepare() { return make_ready_future<>(); }
>     virtual void execute() = 0;
> };
> 
> using external_updater = std::unique_ptr<external_updater_impl>;
> 
> template <std::invocable_r<void> Execute>
> external_updater
> make_simple_external_updater(Execute func) {
>     class thunk : public external_updater_impl {
>          Execute _func;
>     public:
>          virtual void execute() override {
>               _func();
>          }
>     };
>     return std::make_unique<thunk>(std::move(func));
> }
> ```
> 
> We can run the prepare part first, then execute() atomically with the cache thing.

Looks very good! It will improve the current scenario where external updaters aren't really providing strong exception guarantees in the "execution" phase.